### PR TITLE
refactor(rome_formatter): new format API

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

This is part #1996 

This is a first batch of changes where we start using the new traits. While doing so I uncovered two new cases which now implement the new traits.

The changes should be applied to all the nodes but I decided to start small in order to not make the PR too overwhelming.

## Test Plan

No tests should fail

### CODE REVIEW

File: crates/rome_formatter/src/formatter_traits.rs [hunk 1]
```diff
@@ -214,6 +214,27 @@ impl FormatOptionalTokenAndNode for Option<SyntaxToken> {
     }
 }
 
+impl FormatOptionalTokenAndNode for SyntaxResult<Option<SyntaxToken>> {
```
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_formatter/src/formatter_traits.rs:217]: I think you're onto something here!!

[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_formatter/src/formatter_traits.rs:220]: I tried to apply your suggestions and the compiler flags some conflicting implementation. 

For example:
```rs
impl<F: FormatTokenAndNode> FormatOptionalTokenAndNode for Option<F> 
// and
impl<F: FormatOptionalTokenAndNode> FormatOptionalTokenAndNode for Option<F>
```

Although, using the trait bounds, we can actually cover all the cases.

**Comment by MichaReiser** [comment]
!bench_formatter

File: crates/rome_formatter/src/formatter_traits.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leops) [reply crates/rome_formatter/src/formatter_traits.rs]: I think the second one should be `impl<F: FormatTokenAndNode> FormatOptionalTokenAndNode for F {}` (anything that can be formatted can also be "optionally formatted")

[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_formatter/src/formatter_traits.rs]: which overlaps with

```rs
 impl<F: FormatOptionalTokenAndNode> FormatOptionalTokenAndNode for SyntaxResult<F> 
```

[the remark this answers is not in the document]
Comment (leops) [reply crates/rome_formatter/src/formatter_traits.rs]: Ah I see, well that's why I wasn't sure if the type system in Rust would accept it

File: crates/rome_formatter/src/ts/assignment/object_assignment_pattern.rs [hunk 2]
```diff
@@ -51,11 +51,9 @@ impl ToFormatElement for JsAnyObjectAssignmentPatternMember {
...
 impl ToFormatElement for JsObjectAssignmentPatternShorthandProperty {
     fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
-        let init_node = if let Some(node) = self.init() {
-            format_elements![space_token(), formatter.format_node(&node)?]
-        } else {
-            empty_element()
-        };
+        let init_node = self
+            .init()
+            .format_with_or_empty(formatter, |node| format_elements![space_token(), node])?;
```
Comment (NicholasLYang) [inline crates/rome_formatter/src/ts/assignment/object_assignment_pattern.rs:56]: This is a pretty nice api!

File: crates/rome_formatter/src/ts/auxiliary/any_function.rs [hunk 3]
```diff
@@ -11,32 +12,27 @@ impl ToFormatElement for JsAnyFunction {
...
-            tokens.push(formatter.format_token(&token)?);
-        }
+        tokens.push(self.function_token().format_or_empty(formatter)?);
+        tokens.push(self.star_token().format_or_empty(formatter)?);
 
         tokens.push(match self {
             JsAnyFunction::JsArrowFunctionExpression(_) => empty_element(),
             _ => match self.id()? {
-                Some(id) => format_elements![space_token(), formatter.format_node(&id)?],
+                Some(id) => format_elements![space_token(), id.format(formatter)?],
```
Comment (NicholasLYang) [inline crates/rome_formatter/src/ts/auxiliary/any_function.rs:26]: Can't we do:
```
self.id()?.format_with_or(
                formatter,
                |id_token| format_elements![space_token(), id_token],
                space_token,
            )?
```

**Review by NicholasLYang — COMMENTED** [review]
I like this API! Very composable

**Pull request merged**